src/subdataframe/subdataframe.jl

```diff
@@ -176,32 +176,106 @@ Base.@propagate_inbounds Base.getindex(df::SubDataFrame, row_ind::typeof(!),
                                        col_inds::MultiColumnIndex) =
     select(df, col_inds, copycols=false)
 
-
 Base.@propagate_inbounds function Base.setindex!(sdf::SubDataFrame, val::Any, idx::CartesianIndex{2})
-    setindex!(sdf, val, idx[1], idx[2])
+    return setindex!(sdf, val, idx[1], idx[2])
 end
+
 Base.@propagate_inbounds function Base.setindex!(sdf::SubDataFrame, val::Any, ::Colon, colinds::Any)
-    parent(sdf)[rows(sdf), parentcols(index(sdf), colinds)] = val
+    if colinds isa SymbolOrString && columnindex(sdf, colinds) == 0
+        if !is_column_adding_allowed(sdf)
+            throw(ArgumentError("creating new columns in a SubDataFrame that subsets " *
+                                "columns of its parent data frame is disallowed"))
+        end
+        if !(val isa AbstractVector && nrow(sdf) == length(val))
+            throw(ArgumentError("Assigned value must be a vector with length " *
+                                "equal to number of rows in the SubDataFrame"))
+        end
+        T = eltype(val)
+        newcol = Tables.allocatecolumn(Union{T, Missing}, nrow(parent(sdf)))
+        fill!(newcol, missing)
+        newcol[rows(sdf)] = val
+        parent(sdf)[!, colinds] = newcol
+    else
+        parent(sdf)[rows(sdf), parentcols(index(sdf), colinds)] = val
+    end
+    return sdf
+end
+
+function Base.setindex!(sdf::SubDataFrame, v::AbstractVector,
+                        ::typeof(!), col_ind::ColumnIndex)
+    if col_ind isa Union{Signed, Unsigned} && !(1 <= col_ind <= ncol(sdf))
+        throw(ArgumentError("Cannot assign to non-existent column: $col_ind"))
+    end
+    if col_ind isa SymbolOrString && columnindex(sdf, col_ind) == 0
+        if !is_column_adding_allowed(sdf)
+            throw(ArgumentError("creating new columns in a SubDataFrame that subsets " *
+                                "columns of its parent data frame is disallowed"))
+        end
+        sdf[:, col_ind] = v
+    else
+        pdf = parent(sdf)
+        p_col_ind = parentcols(index(sdf), col_ind)
+        old_col = pdf[!, p_col_ind]
+        T = eltype(old_col)
+        S = eltype(v)
+        newcol = Tables.allocatecolumn(promote_type(T, S), length(old_col))
+        newcol .= old_col
```

Comment:
```suggestion
        newcol = similar(old_col, promote_type(T, S))
        copyto!(newcol, old_col)
```

Reply:
Right. The only question is if we should use `promote_type` or `Base.promote_typejoin`. What do you think?

In particular let us discuss a corner case of when `old_col` is a `CategoricalVector` holding e.g. `String` and `v` is a vector of `String`. In this case we have:
```
julia> Base.promote_typejoin(CategoricalValue{String, UInt32}, String)
Any

julia> promote_type(CategoricalValue{String, UInt32}, String)
String
```
so:
1. we get a significantly different result in both options.
2. in BOTH cases the result of `similar` will not be a `CategoricalVector` but just a `Vector`

Reply:
I think `promote_type` is better. We use it in other places IIRC, it's better to get homogeneous types in general. In the CategoricalArrays example, `Vector{String}` is better than `Vector{Any}` (which is why `promote_type` methods are defined that way).

Reply:
OK - I keep `promote_type` then. I will ammend the documentation to reflect this.